### DISCUSSION TITLE
What is a signature?

### DISCUSSION THREAD

**Discussion by radu-matei** [bytecodealliance/SIG-Registries#7] [General]
It is important to distinguish between the following objects (I am open to discuss the names of the objects):

1. named entity: `bytecodealliance:url-parser`
2. entity release: `bytecodealliance:url-parser:1.0.0`
3. entity content: `bytecodealliance-url-parser-1.0.0.wasm`

Which object signatures should the registry be concerned with, and which should be the concern of the consumer?

One view is that the registry should be concerned with signatures for named entities (1) and entity releases (2), while signatures for binary objects (3) are the responsibility of the consumer (see https://github.com/wasm-signatures/design)

What signatures should the registry handle?

**Comment by carlsverre** [comment]
It seems that the registry should be concerned with verifying the integrity of it's index and how that index maps to content. This seems to align with your proposal that the registry should focus on signing named entities and releases rather than the individual objects. This also aligns well with signature reuse between our registry and other ways to distribute and verify wasm objects. Hence why the wasm-signatures proposal is useful and something to work with rather than replace.

That said, there are a couple cases where the registry may want to at least support reading content level signatures:

- indexing and exposing signatures stored in a wasm file through the api (i.e. through the use of [wasm-signatures](https://github.com/wasm-signatures/design))
- allowing signatures to be attached to content which doesn't support embedded signatures (i.e. static assets)
  - alternatively it would be reasonable to make this the responsibility of bindle and consider using that as our primary package format

**Comment by carlsverre** [comment]
On the topic of signatures I just came across https://www.sigstore.dev/ which looks highly aligned with our goals. Should we consider integrating with them rather than rolling our own signature solution? They have an existing signature storage and verification system called https://docs.sigstore.dev/rekor/overview which is pretty nice. In addition it supports the use of openid connect based signatures which can improve ergonomics by allowing a user to use their openid connect based identity to sign without the additional need of managing keys. https://docs.sigstore.dev/cosign/openid_signing

**Reply by lann** [reply]
Yeah, I've been keeping an eye on sigstore. I have some concerns, but it definitely has some appeal in terms of not reinventing that infrastructure.

* `cosign` (signing CLI) is currently somewhat geared toward OCI registries, but it does have generic blob signing and they may be amenable to supporting us
* `rekor` (transparency log) is not yet 1.0, but they claim to be working on a public transparency log, which would be great if it fits our needs
* `fulcio` (OIDC PKI) looks interesting, but I have reservations about the potential reduction in security from tying package signatures to a regular (often password-authed) account, though the flip-side of that is that humans don't have a great track record with key management, so maybe it's net positive. It's also in the same boat as `rekor`: pre-1.0, promised public instance

There are also some sigstore libraries (including `sigstore-rs`) which I would _definitely_ take a look at.

**Reply by carlsverre** [reply]
Cosign seems to have started with containers as the focus but now seems to have pretty good standalone support. See https://github.com/sigstore/cosign/blob/main/FUN.md for examples.

In many cases I would rather leverage OIDC PKI for user trust than self-managed keys. The nice thing about the model is that it's additive, so a registry could decide what it accepts at what level. Maybe organizations need to have properly managed and trusted keys (potentially via some requirement on who/how many trust signatures are needed in certain cases) while individual users can get away with OIDC.

Does anyone have connections into sigstore? Would be nice to chat with them and see if they are working with any other registries.

**Reply by lann** [reply]
I had a brief chat about sigstore with @lukehinds. He suggested we integrate with e.g. [sigstore-rs](https://github.com/sigstore/sigstore-rs) rather than `cosign`, but otherwise I think it looks promising. They're doing a KubeCon EU booth and want to chat more after.

@carlsverre I think I'm convinced on the OIDC PKI. Good point about it not being either/or; I can totally buy (most) individual maintainers using OIDC while (some) orgs maintain their own keys. :+1:

**Reply by guybedford** [reply]
It seems that such ledger systems are based on the assumption of the registry maintaining it's own authority signature system that signs its own internal verifiable logs. The primary benefit of this seems to be that either private auditors or public auditors (depending on whether or not the log is made transparent or not) are able to fully verify the running of the registry by its operators, and that such an audit trail cannot be rewritten to change history. Especially since the signatures interleave with transactions including publisher signatures which can't be forged either.

So I guess the question would be whether having a verifiable audit trail is a property that is deemed desirable or not for the system?

Personally I would rather focus first on the important property of ensuring the property that the entire registry can be easily _forked_ so that if at any point any community deems it in their own interest to run their own instance they can. I wonder how easy that would be with sigstore if the root signatures would need updating in such a scenario?

Just because there's a valid audit trail, doesn't mean the operators aren't using their authority to censor certain types of code in a way that does not align with a given user community.

**Reply by lann** [reply]
sigstore is working to provide public-good instances of rekor (transparency log) and fulcio (OIDC PKI), which I imagine any BCA-operated registry instance would use (if we use sigstore at all). Based on their community meeting today I think those are likely to be GA in the next 2-3 months.

**Reply by guybedford** [reply]
I guess a registry fork could just involve archiving the old transaction log and starting a new one under a new authority, where the index verification would just be the composition of those two logs? Messy territory I know, but it's exactly when considering verifiable ledgers that questions like these verification questions open up.

**Comment by squillace** [comment]
just jotting my thoughts down here about the sigstore toolchain. As with others, these are merely arguments and may involve mistaken assumptions and misunderstanding, so take them with this in mind. It should go without saying, you should all evaluate my comments in the context of being from "the MegaCorp" and therefore maybe more or maybe less persuasive than others might think. (BTW, I'm happy to take further conversation elsewhere, because this area I have tons to learn about like all the others.)

The Sigstore toolchain approach has some issues each of which falls into roughly three categories: community and processes involving trust; implementation issues we must consider involving the core trust anchor, if you will; and how the MegaCorp and Capitalist ecosystsem plays into all this crud. :-)

## The Human Process of Trust
The first problem has to do with how we all establish trust with one another and add others to truly form a community around a project or projects. It's easy to **use** something that someone else has built, and that agility means the temptation to do so is amazing. But that's not really "open" any more than anything else. What we need is the ability to share the creation of something -- precisely what we're doing in this foundation right here. In this SIG, we have all kinds of people from all kinds of orgs, and we have to communicate, agree, and start committing, and that is even true when we start committing on different implementations. That IS our strength here -- and note, I may be from a MegaCorp, but I have no power but argument! :-) 

_Taken collectively_, and that's the important point, sigstore was launched "as a unit" and then a single startup has picked up the mantle -- and if I may they're doing a great job. That doesn't a community movement make, and it would be premature to me to imagine that this is a bet to make _at this time_ at the specification level.  My position is that until sigstore has a free and open community of broad based contributors it's unhealthy to make it "the way" to do something as serious as ensuring secure supply chain. I would vastly prefer if our project specified what needed to happen and then outline the various tools existing in which it could happen (one of which could easily be a tool like cosign or the sigstore toolchain, for example).  That's the open source community health and freedom position.

## Issues Delegating Original Trust 
At an implementation level, the one tool I absolutely LUUUVVVVV in the sigstore chain is the fulcio, which is the tool that does the OIDC dance to establish trust through a hosted service. For the 80% case, this is fine, but there are a few angles in which _even this approach that I love has issues we should think about_. 

The first one is that OIDP is a faulty system, not when it works, but consider the following cases. The first case is simple, and that's when someone else has stolen your creds. Absent MFA, the malicious agent now has escalation privs to inject stuff into your supply chain _which will otherwise be taken as completely valid by the available chain_. Even should you have MFA, the identity chain doesn't and can't know that -- it's implemented by the service provider. The second case might seem rarer, but it's equally important: that oidc address is just good old DNS and those domains get hijacked and reused (this recently just happened). 

These two cases are in the technical regime, but there are ecosystem and ethical reasons to avoid this as well. The first is that certificates are understood and many large organizations trust themselves to run them -- and we have to note, even if this might be wrong to do so. These organizations will not use such a mechanism, full stop. And yet they ARE capable of signing and establishing trusted root identity and so on. They just won't participate in an OIDS mechanism such as that described here. The ethical case is the last one I'll make, but it might be the most important. In sharing things in this world, trust is established by a long lived immutable identity, and public keys perform this function and _not human, state-approved identity mechanisms_. You may not think this is important, but many people do, and so much of the software we are blithely throwing around these days these people built. We must ensure that participating identities do not require any corporate mapping to a personal, economy- or state-approved identity. You don't have to trust a public key, but you can look at the code. Over time, you will trust that key. It "is" them, whoever they are. 

Again, I love fulcio, and it would completely suffice for my needs! But I have always tried to be vigilant and bend over backward to ensure that everyone can build software and use it, not merely the MegaCorps. _And I represent one._

## Capitalism
Heh. This one's fun, because I DO work for a MegaCorp in the US. :-) Individuals working in common to gain trust across social boundaries and financial ones made the world we live it. But because so many companies like mine now fund this work, at the business level it's important to realize that _deliberate balance of funding sources is a primary objective to grow balanced, trustworthy communities_. Here I am expressing a very personal view, and you may vary: I do not trust any single company to fund forever a key project that isn't a broadbased community because I wouldn't trust my own company to do that, either. It is a weakness in any project ecosystem to be primarily funded by one company. I would absolutely _try using it to fulfill implementation requirements_ in order to evaluate it! (I'm a dev after all.) But at the community level in collective work I'd strongly suggest we specify what a thing like a signature requires, and then let anyone choose tools to fulfill that requirement. If the single-megavendor tool becomes a rich and healthy human and business ecosystem, maybe it becomes the default approach. But only at that time should it do so. 

how's that? As you can imagine, I've been noodling about this very specific issue for some time having watch lots of foundations fail, megaprojects and communities fail, and so on. As a result, the very subject, which is about "a signature", should be precisely about THAT and not about which tool provides it until you decide what it is. :-) All that said, you can ignore this, which is really about "sigstore" as an immediate "snap-to" implementation, and let's get back to https://github.com/bytecodealliance/SIG-Registries/discussions/7#discussioncomment-2689464. That's the question to drive. :-) 

##

**Reply by lann** [reply]
I have forked the OIDC PKI discussion into #13.
